### ISSUE TITLE
Run Docker container as non-root user

### ISSUE THREAD

**Issue by robflate** [garethgeorge/backrest#476]
**Is your feature request related to a problem? Please describe.**

I'm running Backrest in Docker. When it creates a repo and subsequent snapshots, all file permissions of the original files are ignored and all files are owned by root when restored. Also, all files in bind mounts like `/data` and `/config` are owned by root. 

**Describe the solution you'd like**

- Pass environment variables `PUID` and `PGID` so bind mounts (`/data` and `/config`) have the same owner as my host.
- Run Backups on files/folders regardless of `USER:GROUP` and permissions.
- Respect original file owner/permissions when restoring files regardless of `USER:GROUP` and permissions.
- Run docker commands without getting permission denied.

It's common for a container to have an entrypoint.sh script that includes something like;

```
PUID=${PUID:-1000}
PGID=${PGID:-1000}

USER=abc
GROUP=abc

if [ "$PUID" == "" -o "$PUID" == "0" ]; then
	USER=root
	GROUP=root
elif ! $(id abc&>/dev/null); then
	groupadd -g $PGID $USER
	useradd -m -u $PUID -g $GROUP $USER
fi
```

**Additional context**

I lack the knowledge to understand how I can both run the container as a specific non-root user whilst also letting the container read/write data owned by different users (including root) that may also have restrictive permissions (`600`). I also don't understand the impact of this on running commands like `docker <container-name> stop` etc. 

Hopefully this is simple without requiring granting any [extended capabilities or privileges](https://docs.docker.com/engine/containers/run/#runtime-privilege-and-linux-capabilities) but I'm afraid that's beyond my understanding.

**Comment by robflate** [comment]
I've updated by post with additional information. It would be great to hear thoughts on this.

**Comment by garethgeorge** [comment]
Thanks for converting this into an FR and the detailed discussion if what you're trying to get done -- I think it makes sense for Backrest to try to adopt the UID= and GID= env vars convention established by linuxservers.io , we can definitely do something like this in a future release. 

This issue might be an easy starter PR if anyone has interest in adding an entrypoint wrapper.

**Comment by robflate** [comment]
I found the Restic docs for running full backups without root. They're hidden away in the examples section;

https://github.com/restic/restic/blob/master/doc/080_examples.rst#full-backup-without-root

It suggests adding the extended attribute `setcap cap_dac_read_search=+ep` to the restic binary. Backrest would still also need to map the local user to the container with `UID/GID` env vars.

**Comment by danktankk** [comment]
has this been implemented yet by chance?

**Comment by garethgeorge** [comment]
It hasn't -- most of the work here is in testing the docker build and validating that permissions work correctly in a variety of contexts.

**Comment by awnumar** [comment]
Just to add some context to this, I believe this is the reason I'm not able to restore any snapshots running the Docker container on Unraid. I get the error 

```
error: restore failed: restore snapshot "...:/backup/shared" for repo ...: command "/bin/nice -n 10 ionice -c 2 -n 7 /bin/restic restore --json ..." failed: exit code 1: restore failed
processing command output: no summary event found
Output:
cannot create target directory: mkdir /backup/shared/.DS_Store-backrest-restore-65d8f743: read-only file system
```

The mounted /backup folder is owned (on the host) by `nobody` in the `users` group (PUID=99, PGID=100)

**Comment by DetermineAbsurd** [comment]
@awnumar If you are using the default docker template from the community addons plugin for unraid, it mounts the backup directory as read only by default. You will have to change the mount to r/w.

**Comment by ajyey** [comment]
Another vote for this. I run backrest in a container and its a pain not being able to view the config or logs without `chown`'ing each time they're written to

**Comment by stefanorg** [comment]
+1 for me. 

@garethgeorge I saw that there is already a feature branch `garethgeorge/uidgid` any plan to add it to main?

**Comment by ciotorcristian** [comment]
I think this feature is very important for the security of our systems. Any news on this?

**Comment by theacodes** [comment]
I'm happy to test this our or help out with a PR if needed, is this still something you're open to adding @garethgeorge ?

**Comment by spacecakes** [comment]
Just ran into this. Can't open the directories I restore because they are owned by root. Is there a fix or workaround?

**Comment by garethgeorge** [comment]
> I'm happy to test this our or help out with a PR if needed, is this still something you're open to adding [@garethgeorge](https://github.com/garethgeorge) ?

Yep -- this is something I'm open to supporting. Started work on this in https://github.com/garethgeorge/backrest/pull/829 but haven't had time to prioritize finishing the impl and validating it. Happy if someone is interested  to pick up the PR (or a better approach :) ).

**Comment by RikudouGoku** [comment]
> Just ran into this. Can't open the directories I restore because they are owned by root. Is there a fix or workaround?

You can use the `user: 99:100` variable and with `UMASK:002` environment variable (on unraid). And the restored files will be owned by the user "nodoby" and with read & write access for owner and group but nothing for "others". 



EDIT: Apparently it does not work IF the backup container ran as root. And the container for restore is non-root.

**Comment by hollowimage** [comment]
so. you can switch the container runtime identity to 99:100 (user:nobody) on unraid even if it initialized as root, you j ust have to `$ chown -R 99:100 ./*` in the appdata the entire bulk of backrest's workset: this will include /config and /data dirs, and logging endpoints -- you can trace down what its missing in container logs for which paths its getting access denied. 

this does make it that something owned by root isnt visible by backrest, even if you map it into the container, but for my purposes, i have a different flow for backup of root-owned components

I am not, however, able to leverage UMASK because that requires wrapping the backrest entrypoint with that umask, and that's natively handled by shell, not go, so you'd have to either change how the container is initialized, or implement this in go, and based on te comment from OCT31, sounds like works's underway, he's just backlogged :) 

edit: one more observation, the container actually alreayd runs with umask 0022, so, there might be some explicit perms being set that are overriding this.

**Comment by HowardRoeske** [comment]
I ended up here after trying to run the Backrest Docker image as dedicated non-root user, and all of the bullet points in the original request perfectly line up with my needs. Currently I'm trying with CAP_DAC_READ_SEARCH granted in the compose file, and am encountering skipped file warnings due to read permission errors (the second bullet point mentioned in the FR).

> I found the Restic docs for running full backups without root. They're hidden away in the examples section;
> 
> https://github.com/restic/restic/blob/master/doc/080_examples.rst#full-backup-without-root
> 
> It suggests adding the extended attribute `setcap cap_dac_read_search=+ep` to the restic binary.

So it looks like the caps need to be available for both the user and the binary.

I'm about to try a work-around by mounting a container volume that contains a restic binary with setcap already applied to it, and directing backrest to it with `BACKREST_RESTIC_COMMAND`.

**Edit:** Yup, that worked!
Here are the relevant bits from my docker-compose file:
```
environment:
  - BACKREST_RESTIC_COMMAND=/restic/bin/restic
  [...]
user: ${ENV_USER}
cap_drop:
  - ALL  # Strip all default root-level capabilities for the container's user
cap_add:
  - DAC_READ_SEARCH  # User can bypass read permission checks (required for running restic with CAP_DAC_READ_SEARCH)
volumes:
  - ./bin:/restic/bin:ro
  [...]
```
Where
- `${ENV_USER}` is the `{user id}:{group id}` of a system-specific low-privilege user for backrest (read in from the `.env file`).
- A copy of the [latest restic binary](https://github.com/restic/restic/releases) is located at `./bin/restic` (which is mounted into the container at `/restic/bin/restic` - the same value passes to `BACKREST_RESTIC_COMMAND`).
- The restic binary is owned by the same user id passed to the container, and has the permissions 100 (only the owner can execute it, and the only thing the owner can do is execute it).

Note that `security_opt: - no-new-privileges:true` cannot be used with this, as it testing indicates it seems to prevent anything being granted via `cap_add`.

**Comment by HowardRoeske** [comment]
> * Run commands inside the container that require elevated privileges (e.g. `docker <container-name> stop`) without getting permission denied. Either run as root or add the mapped user to the Docker group inside the container or by giving the mapped user elevated privileges.

I agree with all of the other points, but this one feels counter to the goal of restricting privileges of the container user, and is pretty specific to the environment you're running (I'm guessing you want to take containers offline during a backup for data integrity issues, and then restart them afterward).

Doing this sort of work as a non-root user seems to be exactly what a Docker socket proxy is for: expose a limited subset of the full Docker functionality that can be invoked by a limited user. Here are a couple of potentials:
- [linuxserver/socket-proxy](https://docs.linuxserver.io/images/docker-socket-proxy/)
- [Tecnativa/docker-socket-proxy](https://github.com/Tecnativa/docker-socket-proxy)

I haven't tried any of these myself, but it feels like it'd address your use case without needing any changes to backrest?

**Comment by HowardRoeske** [comment]
@garethgeorge Would you be interested in a cookbook PR for running backrest as a non-root user, including the method I outlined above for being able to back up all files by using an external copy of restic with the bypass read checks capability set?

It'd be really cool to not require that workaround, but I'm not sure if there's anything in a container that you can do to detect the presence of the `CAP_DAC_READ_SEARCH` capability without requiring `root` to leverage it. I'm fairly new to all of this, though, so there very well may be an approach here that I'm not aware of.

**Comment by HowardRoeske** [comment]
Ran through a bit more rigorous testing, and updated my initial reply (and docker compose example).

**Comment by seb26** [comment]
> Yep -- this is something I'm open to supporting. Started work on this in [#829](https://github.com/garethgeorge/backrest/pull/829) but haven't had time to prioritize finishing the impl and validating it. Happy if someone is interested to pick up the PR (or a better approach :) ).

Thank you for working on this, I gave comments on this PR to go with a bash approach for an entrypoint.

I might propose it as a separate PR to solve this issue.

But I got tied up in the Dockerfile/gorelease structure as it stands at the moment.

If I pushed my patches now it would disregard Dockerfile.alpine and Dockerfile.scratch, reunite them in a single file, use restic from apk instead of backrest get dependencies, and other somewhat big deviations in the build workflow...?
